File: java/src/distjoint_set/randomTestData.java
package distjoint_set;

import java.util.Random;

class randomTestData {
	private static final Random rand = new Random(0);

	int ops;
	int[] indexes;
	boolean[] expectedUnions;
	int[] expectedFinds;

	randomTestData(int n) {
		int[] par, size;
		par = new int[n];
		size = new int[n];
		for (var i = 0; i < n; i++) {
			par[i] = i;
			size[i] = 1;
		}

		ops = n + rand.nextInt(n*2);
		indexes = new int[ops*3];
		expectedUnions = new boolean[ops];
		expectedFinds = new int[ops];

		for (var op = 0; op < ops; op++) {
			int j = rand.nextInt(n), i = rand.nextInt(n), k = rand.nextInt(n);
			indexes[op*3] = j;
			indexes[op*3+1] = i;
			indexes[op*3+2] = k;

			j = find(par, j); i = find(par, i);
			if (j != i) {
				expectedUnions[op] = true;

				if (size[j] >= size[i]) { par[i] = j; size[j] += size[i]; }
				else { par[j] = i; size[i] += size[j]; }
			}
			expectedFinds[op] = find(par, k);
		}
	}

	static int find(int[] par, int i) {
		while (i != par[i]) {
			var p = par[i];
			par[i] = par[p];
			i = p;
		}

		return i;
	}
}

File: go/disjoint_set/random_test_data.go
package disjoint_set

import (
	"math/rand/v2"
)

type randomTestData struct {
	ops            int
	indexes        []int
	expectedUnions []bool
	expectedFinds  []int
}

var random = rand.New(rand.NewPCG(0, 0))

func newRandomTestData(n int) *randomTestData {
	var td randomTestData

	find := func(par []int, i int) int {
		for i != par[i] {
			j := i
			i = par[i]
			par[j] = par[i]
		}
		return i
	}

	par, size := make([]int, n), make([]int, n)
	for i := range n {
		par[i] = i
		size[i] = 1
	}

	td.ops = n + random.IntN(n*2)
	td.indexes = make([]int, td.ops*3)
	td.expectedUnions = make([]bool, td.ops)
	td.expectedFinds = make([]int, td.ops)

	for op := range td.ops {
		i, j, k := random.IntN(n), random.IntN(n), random.IntN(n)
		td.indexes[op*3] = i
		td.indexes[op*3+1] = j
		td.indexes[op*3+2] = k

		i, j = find(par, i), find(par, j)
		if i != j {
			td.expectedUnions[op] = true

			if size[i] > size[j] {
				par[j] = i
				size[i] += size[j]
			} else {
				par[i] = j
				size[j] += size[i]
			}
		}
		td.expectedFinds[op] = find(par, k)
	}

	return &td
}
